src/corelibs/U2Formats/src/BAMUtils.cpp

```diff
@@ -606,27 +631,48 @@ static int bamMergeCore(const QString& outFileName, const QList<QString>& filesT
     return 0;
 }
 
-GUrl BAMUtils::mergeBam(const QStringList& bamUrls, const QString& mergedBamTargetUrl, U2OpStatus& os) {
-    coreLog.details(tr(R"(Merging BAM files: "%1". Resulting merged file is: "%2")")
-                        .arg(QString(bamUrls.join(",")))
-                        .arg(QString(mergedBamTargetUrl)));
+static constexpr int MAX_FILES_OPENED = 100;
+static int recursiveBamMergeCore(const QString& outFileName, const QList<QString>& filesToMerge, QStringList& intermediateFiles) {
+    int size = filesToMerge.size();
+    CHECK(size != 0, -1);
 
-    int rc = bamMergeCore(mergedBamTargetUrl, bamUrls);
-    CHECK_EXT(rc >= 0, os.setError(tr("Failed to merge BAM files: %1 into %2").arg(bamUrls.join(",")).arg(mergedBamTargetUrl)), {});
-    return mergedBamTargetUrl;
+    auto mergeSplit = U2Region::split({ 0, size }, MAX_FILES_OPENED);
+    if (mergeSplit.size() == 1) {
+        return localBamMergeCore(outFileName, filesToMerge);
+    }
+
+    QStringList newOutFileNameList;
+    for (int i = 0; i < mergeSplit.size(); i++) {
+        const auto& currentRange = mergeSplit.at(i);
+         QList<QString> newFilesToMerge = filesToMerge.mid(currentRange.startPos, currentRange.endPos());
+        QString newOutFileName = newFilesToMerge.first();
+        do {
+            // Remove ".bam" from the end
+            newOutFileName = newOutFileName.left(newOutFileName.size() - 4);
+            newOutFileName = newOutFileName + "_" + QString::number(i) + ".bam";
+        } while (QFile(newOutFileName).exists());
+
+        newOutFileNameList << newOutFileName;
+        intermediateFiles << newOutFileName;
+        int res = localBamMergeCore(newOutFileName, newFilesToMerge);
+        CHECK(res >= 0, res);
+
+    }
+    return recursiveBamMergeCore(outFileName, newOutFileNameList, intermediateFiles);
 }
 
-void* BAMUtils::loadIndex(const QString& filePath) {
-    // See bam_index_load_local.
-    QString mode = "rb";
-    NP<FILE> file = BAMUtils::openFile(filePath + ".bai", mode);
-    if (file == nullptr && filePath.endsWith("bam")) {
-        file = BAMUtils::openFile(filePath.chopped(4) + ".bai", mode);
+int BAMUtils::bamMergeCore(const QString& outFileName, const QList<QString>& filesToMerge) {
+    coreLog.trace("bamMergeCore: " + filesToMerge.join(",") + " to " + outFileName);
+
+    QStringList intermediateFiles;
```

Comment:
Intermediate files should be cleaned up right after they merged. In this case the disk space overhead will be <=1x (where x is the size of the result file which is approximately equal to the total size of all input files or a total size of the merged files per recursive iteration).

When you clean them up at the end the overhead is (N-1)x where N is a depth of the recursion. So if a user has 200Gb file you may end up with 10x overhead: 2Tb instead of 200Gb.

Reply:
OK